File: _ETC/B형대비문제/Social media/Social media/정답코드/CPP/fast/2.cpp
//18,692KB	106MS	C++	3,120B
int heapSize = 0;
int maxUser;
const int MAX_TIME = 100001;
int users[1001];
int followed[1001][26];
int followedSize[1001];
int followedStop[1001];
 
int posted[1001][135];
int postedSize[1001];
int comparePrecalc[MAX_TIME];
int post[MAX_TIME];
int postTime[MAX_TIME];
int compare[MAX_TIME];
int data[300];
 
 
void siftDown(register int i) {
    while (2 * i + 1 < heapSize) {
        register int left = (i<<1) + 1;
        register int right = (i<<1) + 2;
        register int j = left;
        if (right < heapSize && compare[data[right]] > compare[data[left]])
            j = right;
        if (compare[data[i]] >= compare[data[j]])
            break;
        register int temp = data[i];
        data[i] = data[j];
        data[j] = temp;
        i = j;
    }
}
 
int extractMin() {
    register int min = data[0];
    data[0] = data[heapSize - 1];
    heapSize--;
    siftDown(0);
    return min;
}
 
void buildHeap() {
    for (register int i = heapSize >>1; i >= 0; i--) {
        siftDown(i);
    }
}
 
void init(int N) {
    for (register int i = 0; i <= N; i++) {
        users[i] = 0;
        followedSize[i] = 1;
        postedSize[i] = 0;
        followed[i][0] = i;     
    }
}
 
void follow(const int uID1, const int uID2, const  int timestamp) {
    followed[uID1][followedSize[uID1]++] = uID2;
}
 
void makePost(const int uID,const  int pID,const int timestamp) {
    posted[uID][postedSize[uID]++] = pID;
    postTime[pID] = timestamp;
    post[pID] = 0;
    comparePrecalc[pID] = timestamp;
}
 
void like(const int pID,const  int timestamp) {
    post[pID]++;
    comparePrecalc[pID] +=MAX_TIME;
}
 
void getFeed(const int uID, const  int timestamp, int pIDList[]) {
    heapSize = 0;
    register int currMin = timestamp;
    for (register int i = 0; i < followedSize[uID]; i++) {
        register int ff = followed[uID][i];
        register int to = postedSize[ff] < 11? 0 : postedSize[ff] - 10;
        followedStop[ff] = -1;
        for (register int j = postedSize[ff]-1; j >= to ; j--) {
            register int i1 = posted[ff][j];
            if(postTime[i1] >= timestamp-1000) {
                if(heapSize > 10){
                    register int currTime = comparePrecalc[i1];// (post[i1]<<12) + postTime[i1];
                    if(currTime <= currMin)
                        continue;
 
                    data[heapSize++] = i1;
                    compare[i1] = currTime;
                }else {
                    data[heapSize++] = i1;
                    compare[i1] = comparePrecalc[i1];//(post[i1]<<12) + postTime[i1];
                    if(currMin > comparePrecalc[i1]){
                        currMin = comparePrecalc[i1];
                    }
                }
            } else{
                followedStop[ff] = j;
                break;
            }
        }
    }
 
    if(heapSize < 10){
        for (register int i = 0; i < followedSize[uID]; i++) {
            register int ff = followed[uID][i];
            register int to = followedStop[ff] < 11? 0 : followedStop[ff] - 10;
            for (register int j =  followedStop[ff]; j >= to ; j--) {
                register int i1 = posted[ff][j];
                if(heapSize > 10){
                    if(postTime[i1] <= currMin)
                        break;
                    data[heapSize++] = i1;
                    compare[i1] = postTime[i1];
                }else {
                    data[heapSize++] = i1;
                    compare[i1] = postTime[i1];
                    if(currMin > compare[i1]){
                        currMin = compare[i1];
                    }
                }
            }
        }
    }
    buildHeap();
    register int to = heapSize< 11? 0: heapSize -10;
    register int resPos =0;
    for (register int i = heapSize-1; i>= to; i--) {
        pIDList[resPos++] = extractMin();
    }
}

File: _ETC/B형대비문제/Social media/Social media/정답코드/Java/fast/2.java
//72,604KB	268MS	JAVA	2,332B
class UserSolution {
 
    private static boolean[][] followers;
    private static int[] iposts = new int[200000];
    private static int[] times = new int[200000];
    private static int[] uids = new int[200000];
    private static int[] likes = new int[200000];
    private static int[] selected = new int[200000];
    private static int nPosts;
 
    public void init(int N) {
        followers = new boolean[1001][1001];
        nPosts = 1;
        for (int i = 1; i < followers.length; i++) {
            followers[i][i] = true;
        }
    }
 
    public void follow(int uID1, int uID2, int timestamp) {
        followers[uID1][uID2] = true;
    }
 
    public void makePost(int uID, int pID, int timestamp) {
 
        likes[pID] = 0;
        iposts[nPosts] = pID;
        uids[nPosts] = uID;
        times[nPosts] = timestamp;
        nPosts++;
    }
 
    public void like(int pID, int timestamp) {
        likes[pID]++;
    }
 
    public void getFeed(int uID, int timestamp, int[] pIDList) {
 
        for (int i = 0; i < pIDList.length; i++) {
            pIDList[i] = 0;
        }
        boolean[] follows = followers[uID];
        int endtime = timestamp - 1000;
        int nSelected = 0;
        int i;
        for (i = nPosts - 1; i >= 0 && times[i] >= endtime; i--) {
            if (follows[uids[i]]) {
                selected[nSelected++] = i;
            }
        }
 
        sort(0, nSelected - 1, selected);
        if (nSelected < 10) {
            for (; i >= 0 && nSelected < 10; i--) {
                if (follows[uids[i]]) {
                    selected[nSelected++] = i;
                }
            }
        }
        int end = nSelected > 10 ? 10 : nSelected;
        for (i = 0; i < end; i++) {
            pIDList[i] = iposts[selected[i]];
        }
    }
 
    private static void sort(int left, int right, int[] list) {
 
        if (left >= right)
            return;
        int pivot = list[left + (right - left) / 2];
        int i = left;
        int j = right;
        while (true) {
            while (compare(pivot, list[i]))
                i++;
            while (compare(list[j], pivot))
                j--;
            if (i >= j)
                break;
            swap(i, j, list);
            i++;
            j--;
        }
        sort(left, j, list);
        sort(j + 1, right, list);
 
    }
 
    private static boolean compare(int nPost1, int nPost2) {
 
        if (likes[iposts[nPost1]] < likes[iposts[nPost2]])
            return true;
        if (likes[iposts[nPost1]] > likes[iposts[nPost2]])
            return false;
 
        if (times[nPost1] < times[nPost2])
            return true;
        if (times[nPost1] > times[nPost2])
            return false;
 
        return false;
    }
 
    private static void swap(int a, int b, int[] list) {
        int temp;
        temp = list[a];
        list[a] = list[b];
        list[b] = temp;
    }
}